File: fetch/generate_stats.php
<?php
$outputDir = __DIR__ . '/../docs/data/';
file_put_contents('php://stderr', "📁 输出目录: " . $outputDir . "\n");
file_put_contents('php://stderr', "📁 脚本目录: " . __DIR__ . "\n");
file_put_contents('php://stderr', "📁 当前工作目录: " . getcwd() . "\n");

$official = [];
$community = [];

if (file_exists($outputDir . 'official.json')) {
    $official = json_decode(file_get_contents($outputDir . 'official.json'), true) ?: [];
}
if (file_exists($outputDir . 'community.json')) {
    $community = json_decode(file_get_contents($outputDir . 'community.json'), true) ?: [];
}

$allLibraries = array_merge($official, $community);
$authors = [];
foreach ($allLibraries as $lib) {
    if (!empty($lib['author'])) {
        $authors[$lib['author']] = ($authors[$lib['author']] ?? 0) + 1;
    }
}
arsort($authors);

$stats = [
    'last_update' => date('Y-m-d H:i:s'),
    'total_libraries' => count($allLibraries),
    'official_count' => count($official),
    'community_count' => count($community),
    'authors' => [
        'total' => count($authors),
        'ranking' => array_slice($authors, 0, 20)
    ]
];

if (count($allLibraries) === 0) {
    file_put_contents('php://stderr', "⚠️ 没有数据文件，生成空统计\n");
}

$existingFile = $outputDir . 'stats.json';
$needsUpdate = true;

if (file_exists($existingFile)) {
    $existingStats = json_decode(file_get_contents($existingFile), true) ?: [];
    $existingData = $existingStats;
    $newData = $stats;
    unset($existingData['last_update'], $newData['last_update']);
    
    if (json_encode($existingData) === json_encode($newData)) {
        $needsUpdate = false;
    }
}

if ($needsUpdate) {
    file_put_contents($existingFile, json_encode($stats, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
    file_put_contents('php://stderr', "✅ 统计信息生成成功，总库数: " . $stats['total_libraries'] . "\n");
} else {
    file_put_contents('php://stderr', "✅ 统计信息无变化，跳过写入\n");
}
?>

File: fetch/generate_stats.py
#!/usr/bin/env python3
"""合并 official + community 数据，生成统计信息 → docs/data/stats.json"""

import json
import os
import sys
from datetime import datetime

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(SCRIPT_DIR, '..', 'docs', 'data')


def load_json(path: str) -> list:
    if not os.path.exists(path):
        return []
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


def normalize_authors(all_libraries: list) -> dict:
    """作者名归一化：trim + 忽略大小写合并，保留首次遇到的写法"""
    canonical_map = {}  # lowercase -> canonical form
    authors = {}

    for lib in all_libraries:
        raw_author = lib.get('author', '')
        if not raw_author:
            continue
        raw = raw_author.strip()
        key = raw.lower()

        if key not in canonical_map:
            canonical_map[key] = raw
            authors[raw] = 0

        canonical = canonical_map[key]
        authors[canonical] = authors.get(canonical, 0) + 1

    # 按数量降序
    return dict(sorted(authors.items(), key=lambda x: -x[1]))


def main():
    official = load_json(DATA_DIR + '/official.json')
    community = load_json(DATA_DIR + '/community.json')

    all_libraries = official + community
    authors = normalize_authors(all_libraries)

    stats = {
        'last_update': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
        'total_libraries': len(all_libraries),
        'official_count': len(official),
        'community_count': len(community),
        'authors': {
            'total': len(authors),
            'ranking': authors,
        },
    }

    if not all_libraries:
        print('⚠️ 没有数据文件，生成空统计', file=sys.stderr)

    os.makedirs(DATA_DIR, exist_ok=True)
    output_path = DATA_DIR + '/stats.json'
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(stats, f, ensure_ascii=False, indent=4)

    print(
        f'✅ 统计信息生成成功，总库数: {stats["total_libraries"]}',
        file=sys.stderr,
    )


if __name__ == '__main__':
    main()
